### PR TITLE
Feature/http errors

### PR DESCRIPTION
**Pull request by bmaschio**
@fmontesi @kicito not sure if the test  is correct

### CODE REVIEW

File: extensions/http/src/main/java/jolie/net/HttpProtocol.java [hunk 1]
```diff
@@ -1388,7 +1381,8 @@ private String cutBeforeQuerystring( String requestPath ) {
 		return requestPath.split( "\\?", 2 )[ 0 ];
 	}
 
-	private void recv_checkReceivingOperation( HttpMessage message, DecodedMessage decodedMessage ) {
+	private Optional< RequestErrorCommMessage > recv_checkReceivingOperation( HttpMessage message,
```
Comment (kicito) [inline extensions/http/src/main/java/jolie/net/HttpProtocol.java:1384]: Are we start to use return error as a value practice? I guess waiting for review from @fmontesi

File: extensions/http/src/main/java/jolie/net/HttpProtocol.java [hunk 2]
```diff
@@ -1400,14 +1394,28 @@ private void recv_checkReceivingOperation( HttpMessage message, DecodedMessage d
...
 				if( decodedMessage.resourcePath == null ) {
 					decodedMessage.resourcePath = "/";
 				}
 			}
+			if( hasOperationSpecificParameter( decodedMessage.operationName, Parameters.TEMPLATE ) ) {
+				return Optional.of( new RequestErrorCommMessage( decodedMessage.id, decodedMessage.operationName,
+					decodedMessage.resourcePath, decodedMessage.value, null,
+					getOperationSpecificStringParameter( decodedMessage.operationName, Parameters.TEMPLATE ) ) );
+			}
+			if( hasOperationSpecificParameter( decodedMessage.operationName, Parameters.TEMPLATE ) ) {
```
Comment (kicito) [inline extensions/http/src/main/java/jolie/net/HttpProtocol.java:1410]: duplicate code?

File: test/extensions/http_template.ol [hunk 3]
```diff
@@ -90,6 +97,16 @@ service Test {
...
             }
 
+
+            scope (addOrder){
+                   install ( default =>
+                        if (addOrder.default!= "TypeMismatch" || !is_defined(response.IOException) ){
+                             throw( TestFailed, "Wrong Exception " )
+                        }
+                        );
+                   addOrder@TestHttpNoTemplate({token="sometoken" ammount = 10.0})(response)
```
Comment (kicito) [inline test/extensions/http_template.ol:107]: amount

File: test/extensions/private/http_template_interface.ol [hunk 4]
```diff
@@ -50,8 +50,8 @@ type AddOrderResponse:void
 interface HttpTemplateInterface{
     RequestResponse:
 
-    getOrders(GetOrdersRequest)(GetOrdersResponse),
-    getOrder(GetOrderRequest)( GetOrderResponse),
-    addOrder(AddOrderRequest)(AddOrderResponse),
+    getOrders(GetOrdersRequest)(GetOrdersResponse) ,
+    getOrder(GetOrderRequest)( GetOrderResponse) ,
+    addOrder(AddOrderRequest)(AddOrderResponse) ,
```
Comment (kicito) [inline test/extensions/private/http_template_interface.ol:55]: discard these changes?

**Review by kicito — COMMENTED** [review]
Great works :)

**Comment by fmontesi** [comment]
@kicito and @bmaschio What's the status on this? :)

**Comment by fmontesi** [comment]
@mwallnoefer Do you think you could rebase/integrate this PR after your latest mods? IIRC it introduces the possibility of returning a fault back in case a client tries to invoke a service on an HTTP resource that is not mapped to anything through a template or other means. It also needs some polishing (e.g., I see a missing license header).

What's your take on it?

I think that in general it might be time to discuss whether we should offer some *easy* way of managing general errors, like invoking some default operation for generating 404 pages etc. Currently it's a bit cumbersome, but maybe it's just me.

@bmaschio @klag @kicito What's your experience with returning nice HTML/CSS pages like '404 not found' from Jolie? Any comments are welcome.

**Comment by mwallnoefer** [comment]
I need to see how difficult the rebase will be.

**Comment by mwallnoefer** [comment]
Okay, rebase done and all tests are passing.

**Comment by fmontesi** [comment]
Fantastic, thanks!

**Pull request merged**